File: AirPurrr/app/src/main/java/com/example/trubul/airpurrr/DetectorHelper.java
package com.example.trubul.airpurrr;

import android.content.Context;
import android.util.Log;

import java.util.Arrays;

import androidx.loader.content.AsyncTaskLoader;

class DetectorHelper {
    private static final String TAG = "DetectorHelper";
    private static DetectorCallback mCallback;


    interface DetectorCallback {
        void setPMValuesDetector(Double[] pmValuesDetector);
    }

    DetectorHelper(DetectorCallback callback) {
        mCallback = callback;
    }

    static Double[] download() {
        String rawData;

        try {
            HttpGetRequest getRequest = new HttpGetRequest();
            rawData = getRequest.doHttpRequest(MainActivity.DETECTOR_URL);

            String[] pmStrings = rawData.split("\n");
            Double[] pmDoubles = new Double[pmStrings.length];

            for (int i = 0; i < pmStrings.length; i++) {
                try {
                    pmDoubles[i] = Double.parseDouble(pmStrings[i]);
                }
                catch (NumberFormatException e) {
                    Log.e(TAG, "downloadPMValues: Number Format Exception " + e.getMessage());
                }
            }

            // Convert results to percentages (to ease handling with auto mode)
            pmDoubles = convertToPercent(pmDoubles);

            Log.d(TAG, "downloadPMValues: PMDOUBLES ARE " + Arrays.toString(pmDoubles));
            mCallback.setPMValuesDetector(pmDoubles);

            return pmDoubles;
        } catch (NullPointerException e) {
            Double[] empty = {0.0, 0.0};
            mCallback.setPMValuesDetector(empty);
            return empty;
        }
    }

    private static Double[] convertToPercent(Double[] pmDoubles) {
        Double[] pmDoublesPerc = new Double[2];

        pmDoublesPerc[0] = 4 * pmDoubles[0];  // PM2.5
        pmDoublesPerc[1] = 2 * pmDoubles[1];  // PM10

        return pmDoublesPerc;
    }

    static class Loader extends AsyncTaskLoader<Double[]> {
        Loader(Context context) {
            super(context);
        }

        @Override
        public Double[] loadInBackground() {
            return download();
        }
    }
}

File: AirPurrr/app/src/main/java/com/example/trubul/airpurrr/DetectorHelper.kt
package com.example.trubul.airpurrr

import android.content.Context
import androidx.loader.content.AsyncTaskLoader

private const val DETECTOR_URL = "http://airpurrr.ga/pm_data.txt"

internal class DetectorHelper {

    class Loader(context: Context) : AsyncTaskLoader<MutableList<Double>>(context) {
        override fun loadInBackground(): MutableList<Double> {
            val helper = DetectorHelper()
            return helper.download()
        }
    }

    fun download(): MutableList<Double> {
        val rawData: String?

        try {
            val getRequest = HttpGetRequest()
            rawData = getRequest.doHttpRequest(DETECTOR_URL)

            val pmStrings = rawData!!.split("\n".toRegex()).dropLastWhile { it.isEmpty() }
            var pmDoubles = mutableListOf<Double>()

            for (i in pmStrings.indices) {
                try {
                    pmDoubles.add(pmStrings[i].toDouble())
                } catch (e: NumberFormatException) {
                }
            }

            // Convert results to percentages (to ease handling with auto mode)
            pmDoubles = convertToPercent(pmDoubles)

            return pmDoubles
        } catch (e: NullPointerException) {
            val empty = mutableListOf(0.0, 0.0)
            return empty
        }

    }

    private fun convertToPercent(pmDoubles: List<Double>): MutableList<Double> {
        val pmDoublesPerc = mutableListOf<Double>()

        pmDoublesPerc.add(4 * pmDoubles[0])  // PM2.5
        pmDoublesPerc.add(2 * pmDoubles[1])  // PM10

        return pmDoublesPerc
    }
}